chain-impl-mockchain/src/vote/manager.rs

```diff
@@ -146,17 +183,52 @@ impl ProposalManager {
         &self,
         identifier: account::Identifier,
         payload: ValidatedPayload,
+        token_distribution: &TokenDistribution<TokenIdentifier>,
     ) -> Result<Self, VoteError> {
         // Part of DDoS protection: do not record a new ballot if the account already voted for this
         // proposal. This protects the system from flooding in a system with cheap/free voting
         // transactions.
         let votes_by_voters = self
             .votes_by_voters
-            .insert(identifier, payload)
+            .insert(identifier.clone(), ())
             .map_err(|_| VoteError::AlreadyVoted)?;
+
+        let tally = if let Some(stake) = token_distribution.get_account(&identifier) {
+            match (self.tally.clone(), payload) {
+                (IncrementalTally::Public(mut result), ValidatedPayload::Public(choice)) => {
+                    result.add_vote(choice, stake)?;
+                    IncrementalTally::Public(result)
+                }
+                (
+                    IncrementalTally::Private(mut encrypted_tally),
+                    ValidatedPayload::Private(ballot),
+                ) => {
+                    encrypted_tally.add(&ballot, stake.0);
+                    IncrementalTally::Private(encrypted_tally)
+                }
+                (IncrementalTally::Public(_), ValidatedPayload::Private(_)) => {
+                    return Err(VoteError::InvalidPayloadType {
+                        received: PayloadType::Private,
+                        expected: PayloadType::Public,
+                    })
+                }
+                (IncrementalTally::Private(_), ValidatedPayload::Public(_)) => {
+                    return Err(VoteError::InvalidPayloadType {
+                        received: PayloadType::Public,
+                        expected: PayloadType::Private,
+                    })
+                }
+                (IncrementalTally::Decrypted(_), _) => {
+                    return Err(VoteError::TallyAlreadyDecrypted);
+                }
+            }
+        } else {
+            self.tally.clone()
```

Comment:
And here I think is a good place for an error. If the stake is zero, we are going to write a no-op transactions which opens up potential for attacks.

Reply:
Good point. Although technically the transactions would only be no-ops if the fees are zero, it's better to just reject them altogether.